doc/decisions/error_message.md

```diff
@@ -0,0 +1,137 @@
+#Error message & handling concept
+
+## Problem
+
+The current error concept has disadvantages in following regards:
+
+- Too verbose error message
+- A lot of redundant errors
+- Hard to manage specification file
+- No senseful way for application developers to use error codes from elektra
+
+This concept should improve useability for end users and developers.
+
+## Constraints
+
+- Error numbers must stay because they are more reliable to check against than strings
+
+## Assumptions
+
+## Considered Alternatives
+
+- Removing the specification file without requiring error numbers
+- Possible variations on what message should be displayed
+- Adding the key of the occurred error to the API which permits reading information from
+  additional metadata such as an error message provided by a specification author.
+  Reason against: The description of the key should already provide such information.
+  Doing it in an extra key would imply redundant information.
+
+## Decision
+
+The error message has the current format:
+
+```
+The command kdb set failed while accessing the key database with the info:
+Sorry, the error (#121) occurred ;(
+Description: validation failed
+Reason: Validation of key "<key>" with string "<value>" failed.
+Ingroup: plugin
+Module: enum
+At: ....../src/plugins/enum/enum.c:218
+Mountpoint: <parentKey>
+Configfile: ...../<file>.25676:1549919217.284067.tmp
+```
+
+The new default error message will look like this:
+
+```
+Sorry, plugin <PLUGIN> issued error #<NR>:
+Validation of key "<key>" with string "<value>" failed.
+```
+
+The error #<NR> will be the color red while <PLUGIN> will be the color blue.
+
+Optionally a third line indicating a solution can be added. Eg. for a permission related error there would be a third line:
+
+```
+Possible Solution: Retry the command as sudo (sudo !!)
+```
+
+To avoid losing information, the user can use the command line argument `-v` (verbose) to show
+`Mountpoint`, `Configfile` in addition to the current error message.
+Furthermore a developer can use the command line argument `-d` (debug)
+to show `At` for debugging purposes.
+
+Warnings will be removed from the specification file as they are not needed there.
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
+
+Unused marked errors will be removed from the specification.
+
+The remaining ~130 errors will be categorizes into logical groups with subgroups.
+Each group and subgroup will receive a range of numbers such as in the HTTP protocol.
+
+- Permanent errors (1-500)
+  - Resource (1-50)
+  - Parsing (51 - 100)
+  - Installation (101 - 150)
+  - Logical (151 - 200)
+- Temporary (501 - 1000)
+  - Conflict (501 - 600)
```

Comment:
Conflict is not really temporary: it is permanent until the conflict is resolved...

Reply:
A few days ago you wrote this:

> I think we can split the categories a bit more:
>...
>* Validation into syntactic and semantic (and maybe even others) 
>* Temporary into conflicts and timeouts...

What exactly should I take now. In one sentence you write I should split it and now you tell me that I should not do it?

Reply:
I do not see the contradiction you see here.

"[split] Temporary into conflicts and timeouts" means that there should be two categories: conflicts and timeouts.

Reply:
I understood it the way that I should make a further subdivision but keep the overall category. I understand it now

Reply:
I am a bit confused concerning "Conflicts" even though when reading through #1473. What kind of errors do you imagine under this category? Can you give some examples?

Reply:
Conflict is one kind of error: when kdbSet() fails because some external modification was done. The solution is that the application needs to do kdbGet() afterwards.

The conflict happens because we are optimistic: we assume we are the only one working with the configuration files. If this assumption does not hold, a conflict occurs.